Vg conf

edmc-ss opened the pull request.

Commit:
cp - initial template for new pkg/prob `confgen`

bschatz-swift commented on confgen/api.go:
```diff
@@ -0,0 +1,34 @@
+// Package confgen provides a mechanism by which a supplied configuration is converted into a
+// set of configuration files to be supplied to ProxyFS (proxyfsd), Samba (smbd et. al.), and
+// NFS (nfsd). In (at least) the case of Samba, multiple configuration files will be produced
+// with a per-IP path.
```
NIT: per VIP path

edmc-ss replied:
You know...that's what an earlier version actually said :-).
But I changed it to just "IP" to generically refer to both the VIP and the PublicIPAddr case (the latter being when the VIP field was left blank).
Make sense?

bschatz-swift replied:
ok but I don't think confgen should be used in the PubicIPAddr case.

bschatz-swift commented on confgen/api.go:
```diff
@@ -0,0 +1,34 @@
+// in the case of the ProxyFS configuration file, all VolumeGroups will be included given that
+// ProxyFS needs to know the mapping of VolumeGroup to Peer. The Cluster.WhoAmI can either already
+// be present in the `confFilePath` or be provided as an element of the `confOverrides` argument
+// in each of the APIs below.
```
I think we should state why a given proxyfsd needs to know the location of all VGs and their mapping to peer.    The real reason is bimodal middleware asks proxyfsd where a given volume is mounted.

edmc-ss replied:
I'll add that - thanks!

bschatz-swift commented on confgen/api.go:
```diff
@@ -0,0 +1,34 @@
+
+// ComputeInitial takes a supplied ConfFile, overlays ConfOverrides, and computes an initial
+// set of configuration files that are used by a per-IPAddr set of Samba instances as well as
+// NFSd & ProxyFS.
```
How does ss-proxyfs-monitord pass the mapping of VGs -> nodes?
Is that in the confOverrides array?
Wouldn't it be easier to pass this as a map?   How do you consume this?

edmc-ss replied:
I think the comment at the top kind of gives an example, no?
You would include in the confOverrides []string a string like "VolumeGroup:whatever.PrimaryPeer=PeerWhateverGUID".

edmc-ss replied:
As for it being a map, that's true for the API but not for the cmdline version. Not sure how better to serialize a "map" for the cmdline :-)

edmc-ss replied:
I should also say that this same mechanism is how the Cluster.WhoAmI is set as well...

bschatz-swift replied:
ok

Commit:
cp - enabled ENV setting of program paths

Commit:
cp - completed support for NFS /etc/exports generation

Commit:
cp - fixed cut&paste errors in _test.go files

Note this check-in requires Golang 1.12 (strings.ReplaceAll() missing)

Commit:
cp - finished additions to file_server.conf et. al.

Note 1: Removed dependence on Golang 1.12
Note 2: Still need to add in Cluster Join info somehow...

Commit:
cp - migrated Env Defaults to confgen/api.go

Commit:
cp - added commenting for ^^^

Commit:
cp - frustrated...gotta rework this...

Commit:
cp - re-worked ENVs

Commit:
cp - start the return of smb_common.conf

Commit:
cp - fixed name of return var from fetchCheckpointContainerHeader()

bschatz-swift committed:
First draft of SMB template

Still need to fill out share definitions based on volumes

bschatz-swift committed:
First draft of SMB support

bschatz-swift committed:
Figured out global variables solution

bschatz-swift committed:
Start of shares template

bschatz-swift committed:
Sample proxyfs config file for testing

bschatz-swift committed:
House cleaning

bschatz-swift committed:
Change fetchVolumeInfo() to also grab SMB related info

Start of grabbing SMB info in fetchVolumeInfo()

bschatz-swift committed:
Populate the per VG SMB information

bschatz-swift committed:
Integrate templates and confgen

First draft - still working on the shares

bschatz-swift committed:
Add share support

bschatz-swift committed:
Create the smb.conf files for the different VGs

bschatz-swift committed:
Populate createSMBConf

bschatz-swift committed:
Make style consistent with rest of file

bschatz-swift committed:
Fix VIPs and audit logging in examples

Update variable names

bschatz-swift committed:
Add audit logging at VG level

bschatz-swift committed:
Update for hardening fixes from controller template

bschatz-swift committed:
Cleanup templates with Valid Users per volume

Also, consolidate code

bschatz-swift committed:
Add volume group settings

bschatz-swift committed:
Add new fields from controller

Commit:
Minor fix to swiftclient chaos testing.

Change swiftclient/object.go to use atomic.AddUint64() when it
increments the global variable objectFetchChunkedPutContextCnt,
which avoids an unlikely race condition that might occur during
Chaos unit tests.

Move the increment inside an if statement so the atomic operation
is performed only when globals.chaosFetchChunkedPutFailureRate > 0.

this fix has been sitting around my local tree for quite awhile
and I wanted to finish it.

Commit:
Change confgen to look elsewhere for its templates.

Change the confgen command to look in several places for
"templates/smb_globals.tmpl" and "templates/smb_shares.tmpl":

    /etc/proxyfsd
    /opt/ss/etc/proxyfsd
    .

this is not really the right solution, but overcomes the immediate problem
that confgen is not going to run when invoked by monitord because the
template files that it wants are not going to be in its bin directory.

possibly these belong in a "lib" directory instead of an "etc" directory.

bschatz-swift committed:
Rename file since it only does templates

bschatz-swift committed:
Move smb.go to template.go

bschatz-swift reviewed (approved):
Time to merge this
